connect_four.py

```diff
@@ -37,11 +36,14 @@ def check_col(self, col: str) -> int:
         else:
             return col_i - unicode_A
 
-    def insert_new_piece(self, col: int, color: str) -> bool:
-        row = self.num_rows - len(self.columns[col])
+    def insert_new_piece(self, col: int, color) -> int:
+        row = (self.num_rows - 1 - len(self.columns[col]))
+        print(f"1, row is {row}, col is {col}")
         self.columns[col].append(color)
+        print("2")
         self.current[row][col] = color
-        return True
+        print("3")
+        return row
```

Comment:
I don't think it's necessary to return row for this api. I think the reason you're returning row is that you're using it as an input for board.check_win(). But board.check_win() can perform the same following computation: 

row = (self.num_rows - 1 - len(self.columns[col]))

So the row is extraneous in that api because it can be derived from col. by introducing row, you are creating the possibility of breaking an invariant (row is consistent with the column's progress). This could lead to your api be misused.

Reply:
So I was debating whether to do that or not. I decided not to bc I figured doing the calculation for row in two different methods would create the potential for bugs, but I can see the argument for it. Can you explain what you mean by breaking an invariant?

Reply:
I think you addressed this in your latest code by putting the calculation in once place! correct me if i'm wrong

Reply:
Forget what I said about breaking an invariant by the way.  What I meant to say is that you don't want to create an API where there is a state that is mismatching( the state being a row and column not matching, as in the row does not correctly represent the column's progress). That can lead to issues down the line.

Reply:
I came across this slide deck a while back that I think is worth skimming through to gather some ideas (it's a bit long so don't feel you have to read through the whole thing but it's worth looking through) :

https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/32713.pdf

It's about how to design a good API, written by a Java expert, but the general principles apply.